server_fn/src/lib.rs

```diff
@@ -50,583 +50,564 @@
-            Encoding::GetCBOR => parse_quote!(GetCBOR),
-        };
-        let expansion: syn::Ident = syn::parse_quote! {
-          Encoding::#option
-        };
-        tokens.append(expansion);
+    /// Runs the server function (on the server), bubbling up an `Err(_)` after any stage.
+    #[doc(hidden)]
+    fn execute_on_server(
+        req: Self::ServerRequest,
+    ) -> impl Future<
+        Output = Result<Self::ServerResponse, ServerFnError<Self::Error>>,
+    > + Send {
+        async {
+            let this = Self::from_req(req).await?;
+            let output = this.run_body().await?;
+            let res = output.into_res().await?;
+            Ok(res)
+        }
     }
 }
 
-/// Defines a "server function." A server function can be called from the server or the client,
-/// but the body of its code will only be run on the server, i.e., if a crate feature `ssr` (server-side-rendering) is enabled.
-///
-/// Server functions are created using the `server` macro.
-///
-/// The set of server functions can be queried on the server for routing purposes by calling [`server_fn_by_path`].
-///
-/// Technically, the trait is implemented on a type that describes the server function's arguments.
-pub trait ServerFn<T: 'static>
-where
-    Self: Serialize + DeserializeOwned + Sized + 'static,
-{
-    /// The return type of the function.
-    type Output: serde::Serialize;
-
-    /// URL prefix that should be prepended by the client to the generated URL.
-    fn prefix() -> &'static str;
-
-    /// The path at which the server function can be reached on the server.
-    fn url() -> &'static str;
+#[cfg(feature = "ssr")]
+#[doc(hidden)]
+pub use inventory;
 
-    /// The path at which the server function can be reached on the server.
-    fn encoding() -> Encoding;
+/// Uses the `inventory` crate to initialize a map between paths and server functions.
+#[macro_export]
+macro_rules! initialize_server_fn_map {
+    ($req:ty, $res:ty) => {
+        once_cell::sync::Lazy::new(|| {
+            $crate::inventory::iter::<ServerFnTraitObj<$req, $res>>
+                .into_iter()
+                .map(|obj| (obj.path(), obj.clone()))
+                .collect()
+        })
+    };
+}
 
-    /// Runs the function on the server.
-    #[cfg(any(feature = "ssr", doc))]
-    fn call_fn(
-        self,
-        cx: T,
-    ) -> Pin<Box<dyn Future<Output = Result<Self::Output, ServerFnError>>>>;
+/// A list of middlewares that can be applied to a server function.
+pub type MiddlewareSet<Req, Res> = Vec<Arc<dyn Layer<Req, Res>>>;
+
+/// A trait object that allows multiple server functions that take the same
+/// request type and return the same response type to be gathered into a single
+/// collection.
+pub struct ServerFnTraitObj<Req, Res> {
+    path: &'static str,
+    method: Method,
+    handler: fn(Req) -> Pin<Box<dyn Future<Output = Res> + Send>>,
+    middleware: fn() -> MiddlewareSet<Req, Res>,
```

Comment:
export this? 

```rust
pub middleware
```

Reply:
Ok, now exposing all these fields via methods in 9855bc5. That's what you mean, right?